src/UDPVideo.hpp

```diff
@@ -0,0 +1,151 @@
+//
+//  UDPVideo.hpp
+//  
+//
+//  Created by gene on 9/1/20.
+//
+
+#ifndef UDPVideo_hpp
+#define UDPVideo_hpp
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/types.h> 
+#include <sys/socket.h> 
+#include <arpa/inet.h> 
+#include <netinet/in.h>
+#include <string>
+#include <vector>
+#include <thread>
+#include <mutex>
+#include <iostream>
+#include <queue>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+extern "C" {
+#include <libavutil/avutil.h>
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+#include <libavutil/pixfmt.h>
+}
+
+    struct Image
+    {
+      unsigned char * data; //1280x720x3 MAX out
+      int width;
+      int height;
+    };
+  
+ struct FrameData
+ {
+   unsigned char * data;
+   unsigned int size;
+ };
+
+   class VideoFrameProcessor 
+  {
+   static Image videoBuffer[2]; // Mod by Jeremy
+   static long videoTimestamp[2]; // Mod by Jeremy
+    bool enableCbcuiVision; // Mod by Jeremy
+    static std::queue<AVPacket> FrameDataVector; 
+    static std::queue<AVFrame&> FrameOutput;
+    static std::mutex VidMutex;
+    static std::mutex VidFrameMutex;
+    static bool waitingForReceiver;
+    static int ldestw, ldesth;
+
+	static AVCodecParserContext  *parser;
+	AVDictionary *AVDictH264 = NULL;
+        static AVCodecContext * pCodecCtxH264;
+
+	static int sockfd;
+   static struct sockaddr_in cliaddr;
+   struct sockaddr_in servaddr;
+
+    std::string myDroneAddress;
+    short unsigned int myDronePort;
+
+
+    public:
+
+    VideoFrameProcessor(const char* DroneIpAddress, const short unsigned int DronePort, const int destw, const int desth);
+  
+    //VideoFrameProcessor(const int destw, const int desth);
+    ~VideoFrameProcessor(void) throw();
+    
+    static void addFrametoList(AVPacket packet);
+    void receiverDone(void);
+
+    static bool getpFramefromList (cv::OutputArray image); //AVFrame& newFrame);
+    static void addpFrametoList(AVFrame* newFrame);
+
+     static void run_VFP();	
+     static void run_VDR();
+
+   private:
+	std::thread* VFP_thread;
+    std::thread* VDR_thread;
+
+	static bool VFPrunning;
```

Comment:
why are these static? Maybe this entire class should be a singleton?

Reply:
This class probably should be a singleton, if that means there should be only one of this class.

I'm not too experienced in dealing with singletons (like never). Suggestions or help would be appreciated

Reply:
Would recommend making the constructor `private` so it can't be called by external code. Then create a function kind of like this (untested):

```cpp
public:
typedef std::shared_ptr<ClassName> Ptr;

static Ptr getPtr();

private:
  // private so only we can call
  ClassName();

  // Use a unique_ptr so the memory is automatically released on app termination.
  static const Ptr &g_instance;
  // getRef is not a re-entrant function. If multiple threads race to initialize we could have undefined behavior.
  static std::mutex g_instance_mut;
```

```cpp

ClassName::Ptr ClassName::g_instance = nullptr;
std::mutex ClassName::g_instance_mut;

const ClassName::Ptr &ClassName::getPtr()
{
  std::lock_guard<std::mutex> lock(g_instance_mut);
  if (!g_instance) g_instance = std::shared_ptr<ClassName>(new ClassName);
  return g_instance;
}
```

```cpp
// Usage
const auto &className = ClassName::getPtr();
className->methodCall();
````

Reply:
got rid of the static, they were an artifact of an earlier version.